python_interface/design/04_pyHorace_packaging.md

```diff
@@ -0,0 +1,61 @@
+# A description of the prototype pyHorace implementation
+**2021-1-3**
+
+
+# Background
+
+`pyHorace` will be a Python module which bundles together PACE programs written in Matlab for use by Python users without a Matlab license.
+
+This document describes the options for packaging it for distribution to users.
+
+
+# pyHorace packaging options
+
+There are a few ways to distribute `pyHorace` which we need to decide on:
+
+1. OS-specific packaging system
+2. The Python Package Index (PyPI)
+3. The Anaconda (`conda`) system
+
+Option 1 would be to create packages for specific OS (e.g. `msi` for Windows, `dmg` for Mac OS, `deb` for Debian/Ubuntu, `rpm` for Red Hat/CentOS etc.).
+These packages would be installed by users as they would a non-Python program.
+Options 2 and 3 are specifically for Python packages (or packages meant to be used with Python).
+
+A further decision is whether we should "bundle" the MCR with the `pyHorace` distribution or not.
```

Comment:
Is distribution of the MCR allowed under the terms of the MATLAB licences? When I last looked into this (in around 2012) we concluded that the end user would need to download and install the MCR themselves.

Reply:
This is not entirely clear. If you use the `deploytool` to create a distribution, there is an option to include the MCR with it (making a huge file). The [`MATLAB RUNTIME LICENSE`](https://raw.githubusercontent.com/teknowledgist/Chocolatey-packages/master/manual/mcr-r2019a/MCR_license.txt) itself doesn't say one way or the other whether it is allowed or not (it doesn't explicitly allow or forbid it). I think that given the use in the `deploytool` and the principle that if it is not expressly forbidden, then it is permitted then we can bundle it... but IANAL.

Personally I think that we should *not* bundle the MCR (it will result in huge package artefacts and vastly increase our storage needs).

Reply:
I tend to agree -- unless there is a compelling need to it will be much simpler to inform end-users how to get a compatible version. The only issue that may arise is if we, for some reason, are locked to an old version of the MCR which becomes hard to obtain. I would hope that won't happen and is a problem for another day.

Reply:
Mathworks [provides runtime back to 2012a](https://uk.mathworks.com/products/compiler/matlab-runtime.html) so I think it should be ok. I would suggest that we stick to a single MCR version for longer than the 6 months Matlab release cycle, since we often don't need the new features, and it would mean that users would not have to update the MCR as often. For now I would suggest to stick to R2019b as this is setup on our CI servers and has all the features we need (especially support for C-style row major memory layouts), and to only update the version every 2-3 years (perhaps with every major version change of Horace).

Reply:
Long time now... but this can now be resolved by using the [`compiler.package.installer`](https://uk.mathworks.com/help/compiler/compiler.package.installer.html) command to create a installable package which will download only the components of the MCR needed directly from Mathworks. This saves us bundling the MCR and also presents users with the "Do you accept this license" screen.